File: src/main/scala/com/dylowen/tinglebot/brain/WeightedSet.java
package com.dylowen.tinglebot.brain;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.Serializable;
import java.util.Random;

import gnu.trove.iterator.TObjectIntIterator;
import gnu.trove.map.hash.TObjectIntHashMap;
import gnu.trove.procedure.TObjectIntProcedure;

/**
 * TODO add description
 *
 * @author dylan.owen
 * @since Feb-2016
 */
class WeightedSet<T>
    implements Serializable {
    private final TObjectIntHashMap<T> map = new TObjectIntHashMap<>();
    private int total = 0;

    private transient Random rand = new Random();

    public void add(final T obj) {
        add(obj, 1);
    }

    public void add(final T obj, final int inc) {
        this.map.adjustOrPutValue(obj, inc, inc);

        this.total += inc;
    }

    public T get() {
        TObjectIntIterator<T> iterator = this.map.iterator();
        int offset = 0;

        for (int i = this.map.size(); i > 0; i--) {
            iterator.advance();

            final int rand = this.rand.nextInt(this.total) + 1;
            offset += iterator.value();
            if (rand <= offset) {
                return iterator.key();
            }
        }

        throw new AssertionError("we should never get here");
    }

    public void forEachEntry(TObjectIntProcedure<T> procedure) {
        this.map.forEachEntry(procedure);
    }

    public int size() {
        return this.map.size();
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder();

        TObjectIntIterator<T> iterator = this.map.iterator();

        for (int i = this.map.size(); i > 0; i--) {
            iterator.advance();

            sb.append(iterator.key());
            sb.append(":");
            sb.append(iterator.value());
            sb.append(" ");
        }

        return sb.toString();
    }

    private void readObject(final ObjectInputStream stream)
            throws IOException, ClassNotFoundException {
        stream.defaultReadObject();
        this.rand = new Random();
    }
}

File: src/main/scala/com/dylowen/tinglebot/brain/WeightedSet.scala
package com.dylowen.tinglebot.brain

import java.io.ObjectInputStream

import scala.util.Random

/**
  * TODO add description
  *
  * @author dylan.owen
  * @since Mar-2016
  */
class WeightedSet[T](var map: Map[T, Integer] = Map[T, Integer]()) extends Serializable {
  //private val map: TObjectIntHashMap[T] = new TObjectIntHashMap[T]()
  private var total: Integer = 0

  @transient
  private var rand: Random = new Random()

  def add(obj: T): Unit = add(obj, 1)

  def add(obj: T, inc: Integer): Unit = {
    val weight: Integer = map.getOrElse(obj, 0)

    map += (obj -> (weight + inc))
    //this.map.adjustOrPutValue(obj, inc, inc)

    this.total += inc
  }

  def get: T = {
    var offset: Integer = 0
    val randValue: Integer = this.rand.nextInt(this.total)

    this.map.foreach(entry => {
      offset += entry._2

      if (randValue < offset) {
        return entry._1
      }
    })

    throw new AssertionError("How did you get here? offset = " + offset + " randValue = " + randValue)
  }

  def foreach(f: ((T, Integer)) => Unit): Unit = this.map.foreach(f)

  def size = map.size

  override def toString: String = {
    val builder = StringBuilder.newBuilder

    builder.append(map.head._1)
    builder.append(':')
    builder.append(map.head._2)

    for ((key, weight) <- map.tail) {
      builder.append(" ")
      builder.append(key)
      builder.append(':')
      builder.append(weight)
    }

    builder.toString()
  }

  private def readObject(stream: ObjectInputStream): Unit = {
    stream.defaultReadObject()
    this.rand = new Random()
  }

}
